File: Graphs/1615. Maximal Network Rank.py
class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        g = collections.defaultdict(set)
        for a, b in roads:
            g[a].add(b)
            g[b].add(a)
        res = 0
        visited = set()
        # O(n^2)

        for i in range(n):
            for j in range(n):
                if i != j:
                    t = 0
                    a = len(g[i])
                    b = len(g[j])

                    if i in g[j]:
                        res = max(res, a + b - 1)
                    else:
                        res = max(res, a + b)
        return res

File: Graphs/1615. Maximal Network Rank.cpp
class Solution {
public:
    int maximalNetworkRank(int n, vector<vector<int>>& roads) {
        vector<int>cnt(n, 0);
        vector<vector<int>>connected(n, vector<int>(n, 0));
        for (auto& r : roads) {
            cnt[r[0]]++;
            cnt[r[1]]++;
            connected[r[0]][r[1]] = -1, connected[r[1]][r[0]] = -1;
        }
        int res = 0;
        for (int i = 0; i < n ; i++) {
            for (int j = 0 ; j < n;j++) {
                if (i != j) {
                    res = max(res, cnt[i] + cnt[j] + connected[i][j]);
                }
            }
        }
        return res;



    }
};